### PR TITLE
Reduce transition namespace assignment

### PR DESCRIPTION
**Pull request by jromero-nflx**
Limit transition namespace assignment to unique trunk-eni+sg combinations, instead of creating one per branch-eni association

### CODE REVIEW

**Review by hechaoli — CHANGES_REQUESTED** [review]

File: vpc/service/generate_assignment_id.go [hunk 1]
```diff
@@ -618,36 +618,50 @@ func finishPopulateAssignmentUsingAlreadyAttachedENI(ctx context.Context, req ge
...
-		err = row.Scan(&ass.transitionAssignmentID, &ipv4addr)
+		row := fastTx.QueryRowContext(ctx, `
+			SELECT ass.id
+			FROM branch_eni_attachments bea
+			JOIN assignments ass on ass.branch_eni_association = bea.association_id
+			JOIN branch_enis be on bea.branch_eni = be.branch_eni
+			WHERE is_transition_assignment = true
+			  AND bea.trunk_eni = $1
+			  AND be.security_groups = $2`, ass.trunk, ass.securityGroups)
+		err := row.Scan(&ass.transitionAssignmentID)
```
[the remark this answers is not in the document]
Comment (jromero-nflx) [reply vpc/service/generate_assignment_id.go]: Implemented

File: vpc/service/generate_assignment_id.go [hunk 2]
```diff
@@ -618,36 +618,50 @@ func finishPopulateAssignmentUsingAlreadyAttachedENI(ctx context.Context, req ge
...
+
 		tid.Valid = true
 		tid.Int64 = int64(ass.transitionAssignmentID)
 	}
 
 	// We do this "trick", where we return the values in order to allow a trigger to change the values on write time
 	// for A/B tests.
 	row := fastTx.QueryRowContext(ctx,
-		"INSERT INTO assignments(branch_eni_association, assignment_id, jumbo, bandwidth, ceil, transition_assignment) VALUES ($1, $2, $3, $4, $5, $6) RETURNING id, jumbo, bandwidth, ceil",
+		"INSERT INTO assignments(branch_eni_association, assignment_id, jumbo, bandwidth, ceil, transition_assignment) VALUES ($2, $2, $3, $4, $5, $6) RETURNING id, jumbo, bandwidth, ceil",
```
[the remark this answers is not in the document]
Comment (jromero-nflx) [reply vpc/service/generate_assignment_id.go]: Fixed

File: vpc/service/generate_assignment_id.go [hunk 3]
```diff
@@ -618,36 +618,50 @@ func finishPopulateAssignmentUsingAlreadyAttachedENI(ctx context.Context, req ge
...
-		if err != nil {
-			err = errors.Wrap(err, "Cannot insert transition assignment")
-			tracehelpers.SetStatus(err, span)
-			return err
-		}
 
-		// This should never error because we just did an insert above that should be a noop.
-		row := fastTx.QueryRowContext(ctx, "SELECT id, ipv4addr FROM assignments WHERE is_transition_assignment = true AND branch_eni_association = $1", ass.branch.AssociationID)
-		err = row.Scan(&ass.transitionAssignmentID, &ipv4addr)
+		row := fastTx.QueryRowContext(ctx, `
```
[the remark this answers is not in the document]
Comment (jromero-nflx) [reply vpc/service/generate_assignment_id.go]: Implemented

[the remark this answers is not in the document]
Comment (jromero-nflx) [reply vpc/service/generate_assignment_id.go]: Killswitch implemented

File: vpc/service/generate_assignment_id.go [hunk 4]
```diff
@@ -618,36 +618,50 @@ func finishPopulateAssignmentUsingAlreadyAttachedENI(ctx context.Context, req ge
...
+			WHERE is_transition_assignment = true
+			  AND bea.trunk_eni = $1
+			  AND be.security_groups = $2`, ass.trunk, ass.securityGroups)
+		err := row.Scan(&ass.transitionAssignmentID)
+
 		if err != nil {
-			err = errors.Wrap(err, "Cannot select ID from transition assignments")
-			tracehelpers.SetStatus(err, span)
-			return err
+			transitionAssignmentName := fmt.Sprintf("t-%s", uuid.New().String())
```
[the remark this answers is not in the document]
Comment (jromero-nflx) [reply vpc/service/generate_assignment_id.go:666]: Implemented

File: vpc/service/generate_assignment_id.go [hunk 5]
```diff
@@ -618,36 +618,50 @@ func finishPopulateAssignmentUsingAlreadyAttachedENI(ctx context.Context, req ge
...
+				DO UPDATE SET transition_last_used = now(), gc_tombstone = NULL`,
+				ass.branch.AssociationID, transitionAssignmentName)
+			if err != nil {
+				err = errors.Wrap(err, "Cannot insert transition assignment")
+				tracehelpers.SetStatus(err, span)
+				return err
+			}
+
+			// This should never error because we just did an insert above that should be a noop.
+			row := fastTx.QueryRowContext(ctx, "SELECT id, ipv4addr FROM assignments WHERE is_transition_assignment = true AND branch_eni_association = $1", ass.branch.AssociationID)
```
[the remark this answers is not in the document]
Comment (jromero-nflx) [reply vpc/service/generate_assignment_id.go]: Removed unused var

File: vpc/service/generate_assignment_id.go [hunk 6]
```diff
@@ -870,3 +924,19 @@ LIMIT 1`, req.subnet.SubnetID, req.trunkENIAccount, pq.Array(req.securityGroups)
 	tracehelpers.SetStatus(err, span)
 	return nil, err
 }
+
+func generateTransitionAssignmentName(ctx context.Context, ass *assignment) (*string, error) {
+	_, span := trace.StartSpan(ctx, "generateTransitionAssignmentName")
+	defer span.End()
+
+	bv := []byte(ass.trunk + strings.Join(ass.securityGroups, ""))
```
[the remark this answers is not in the document]
Comment (jromero-nflx) [reply vpc/service/generate_assignment_id.go]: Addressed

File: vpc/service/generate_assignment_id.go [hunk 7]
```diff
@@ -617,29 +620,80 @@ func finishPopulateAssignmentUsingAlreadyAttachedENI(ctx context.Context, req ge
...
+					err = errors.Wrap(err, "Cannot select ID from transition assignments")
+					tracehelpers.SetStatus(err, span)
+					return err
+				}
+			} else {
+				err = errors.Wrap(err, fmt.Sprintf("Cannot select ID from transition assignments for trunk: %s, security groups: %s\n", ass.trunk, ass.securityGroups))
+				tracehelpers.SetStatus(err, span)
+				return err
+			}
+			span.AddAttributes(trace.StringAttribute("transitionAssignmentsExistenceChecked", ass.trunk+"/"+strings.Join(ass.securityGroups, "")))
```
[the remark this answers is not in the document]
Comment (jromero-nflx) [reply vpc/service/generate_assignment_id.go]: Addressed

**Review by hechaoli — APPROVED** [review]
LGTM

**Pull request merged**